crates/polars-io/src/parquet/async_impl.rs

```diff
@@ -142,6 +97,76 @@ impl ParquetObjectStore {
     }
 }
 
+fn read_n<const N: usize>(reader: &mut &[u8]) -> Option<[u8; N]> {
+    if N <= reader.len() {
+        let (head, tail) = reader.split_at(N);
+        *reader = tail;
+        Some(head.try_into().unwrap())
+    } else {
+        None
+    }
+}
+
+fn read_i32le(reader: &mut &[u8]) -> Option<i32> {
+    read_n(reader).map(i32::from_le_bytes)
+}
+
+/// Asynchronously reads the files' metadata
+pub async fn fetch_metadata(
```

Comment:
FWIW https://docs.rs/parquet/latest/parquet/arrow/async_reader/struct.MetadataLoader.html provides an implementation of this which also handles prefetch and loading page indices

Reply:
Thanks for the tip @tustvold. @mickvangelderen can you take a look at that?

Reply:
From what I can see, that implementation is similar. It allows passing a footer size hint to hopefully save one round trip and has tests that also check the number of fetches.

@ritchie46 what do we want to do with that?

Reply:
> @ritchie46 what do we want to do with that?

In what sense? The footer size or the tests? I would favor us the be conservatively here. Rather a larger request but have one, then needing two calls. But that's my gut feeling.

Reply:
For that to make sense we need to have a good idea about how large the footer metadata is in the common case. We are making a trade-off between latency and throughput. If we request a large amount of data so that the footer metadata is likely contained within it, we are possibly waiting to download data that we are going to throw away again (the part that we are downloading but do not use). If we request too few bytes we are affected by both throughput and latency. If we request exactly enough bytes we are affected by latency but not the throughput for bytes that we throw away.

What is most efficient depends on the latency, throughput and footer byte size.

Having to read files backwards is not great. It would be nice if file formats optimize for reading rather than writing.

Anyway, I am happy to reimplement the optimistic fetch. What should the pre-fetch size be? Should it be configurable?